### PR TITLE
Added troubleshooting guide for slow builds

### PR DESCRIPTION
**Pull request by gnufred**
* Added new guide
* Moved stuck build before slow builds guide

### CODE REVIEW

File: src/development/troubleshoot.md [hunk 1]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+## Slow or failing build (or deployment)
+
+Build that take long time or fail is a common problem. Most of the time it's related to application issue and they can be hard to troubleshoot without guidance.
+
+Here are a few tips that can help you solve the issues you are experiencing.
+
+### Check for errors in the logs
+
+Invisible errors during the build and deploy phase can cause increase wait times, failed builds and other problems. Investigating each log and fixing errors is essential.
```
Comment (ralt) [inline src/development/troubleshoot.md]: "increased"

File: src/development/troubleshoot.md [hunk 2]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Related documentation: [Accessing logs](https://docs.platform.sh/development/logs.html#accessing-logs)
+
+### Check php.access.log
+
+The php.access.log contains the page executions by php-fpm. It also includes the execution time and peak memory usage of each request. This can be used to find pages and scripts that use to much memory or to much time to finish.
```
Comment (ralt) [inline src/development/troubleshoot.md]: "too much" x2

File: src/development/troubleshoot.md [hunk 3]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Show 10 slowest page loads in the last 1000 requests: `tail -n 1000 php.access.log | sort -n -k 4 | tail`
+
+### Build and deploy hooks
+
+Hooks are almost always the cause of long build time. If they run into problem they can cause the build to fail or hang indefinitely.
+
+You should test each hook and deploy command in your container (via SSH) with the command `php /app/your_public_folder/your_script.php`.
```
Comment (ralt) [inline src/development/troubleshoot.md]: It's not really possible to test the build hook in the container, so change the wording a bit there?

File: src/development/troubleshoot.md [hunk 4]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Furthermore, you can test your hooks with those linux commands to help figure out any problems
+`time $cmd` # Print execution time
```
Comment (ralt) [inline src/development/troubleshoot.md]: You'll want triple backticks there, to make a proper code block.

File: src/development/troubleshoot.md [hunk 5]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+`strace -T $cmd` # Print a system call report
+
+Ps.: Only use these commands in hooks or scripts if you redirect the output to a file by appending ` 2>&1 > strace.log` to your command.
```
Comment (ralt) [inline src/development/troubleshoot.md]: Proper usage is: `PS: Only...` afaik.

Comment (ralt) [inline src/development/troubleshoot.md]: The hook outputs goes to either git push output or deploy.log file, so I'm not sure why you want to redirect that way. Besides, `strace` has a `-o` option.

File: src/development/troubleshoot.md [hunk 6]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Related documentation: [Build and deploy hooks](https://docs.platform.sh/configuration/app/build.html#hooks)
+
+### Cron jobs
+
+In order to switch containers, we need to wait for all the current running tasks to finish. Long running or failing cron jobs will prevent the container to switch.
```
Comment (ralt) [inline src/development/troubleshoot.md]: "Failing"?

Comment (ralt) [inline src/development/troubleshoot.md]: Also, it'd be more English to say "will prevent the container from switching", no?

File: src/development/troubleshoot.md [hunk 7]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+First, make sure your custom cron jobs’ execution times are low and that they are running properly. Second, investigate your application cron jobs, as they can invoke other services in unexpected ways, which may increase execution time.
```
Comment (ralt) [inline src/development/troubleshoot.md]: Is that a weird apostrophe? :stuck_out_tongue:

File: src/development/troubleshoot.md [hunk 8]
```diff
@@ -161,3 +135,72 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+**About Drupal**: `drush core-cron` run insalled module’s cron task. Those can be, for example; evicting invalid cache, updating database records, regenerating assets. Be sure to constantly benchmark the `drush core-cron` command in all your environments, as it is a common source of performance issues.
```
Comment (ralt) [inline src/development/troubleshoot.md]: `s/insalled/installed/`

Comment (ralt) [inline src/development/troubleshoot.md]: This should probably be a note? Grep for "note" in the docs, there's some markup for that iirc.

File: src/development/troubleshoot.md (hunk 5 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: This was added by @inominate in the first document review. I would honestly just remove it.

File: src/development/troubleshoot.md [hunk 9]
```diff
@@ -161,3 +135,74 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+When a _deployment_ is blocked, you should try the following:
+
+1. Use [SSH](/development/access-site.md) to connect to your environment. Find any long-running cron jobs on the environment by running `ps afx`. Once you have identified the long running process on the environment, kill it.
+2. If you're performing "Sync", "Merge", or "Activate" on an environment and the process is stuck, use [SSH](/development/access-site.md) to connect to the parent environment and identify any long running cron jobs with `ps afx`. Kill the job(s) if you see any.
+3. If you couldn't find any long running process on your environment or parent environment, or you cannot use SSH to connect to your environment, please create a [support ticket](https://platform.sh/support).
+
+
+## Slow or failing build (or deployment)
+
+Build that take long time or fail is a common problem. Most of the time it's related to application issue and they can be hard to troubleshoot without guidance.
```
Comment (ralt) [inline src/development/troubleshoot.md]: "Builds"
"to an application issue"

File: src/development/troubleshoot.md [hunk 10]
```diff
@@ -161,3 +135,74 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+### Check for errors in the logs
+
+Invisible errors during the build and deploy phase can cause increased wait times, failed builds and other problems. Investigating each log and fixing errors is essential.
+
+Related documentation: [Accessing logs](https://docs.platform.sh/development/logs.html#accessing-logs)
+
+### Check php.access.log
+
+The php.access.log contains the page executions by php-fpm. It also includes the execution time and peak memory usage of each request. This can be used to find pages and scripts that use to much memory or time to finish.
```
Comment (ralt) [inline src/development/troubleshoot.md]: "too much", not "to much" :slightly_smiling_face:

File: src/development/troubleshoot.md [hunk 11]
```diff
@@ -161,3 +135,74 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Show 10 slowest page loads in the last 1000 requests: `tail -n 1000 php.access.log | sort -n -k 4 | tail`
+
+### Build and deploy hooks
+
+Hooks are almost always the cause of long build time. If they run into problem they can cause the build to fail or hang indefinitely.
+
+You should test each hook and deploy command in your environment (via SSH) with the command `php /app/your_public_folder/your_script.php`.
```
Comment (ralt) [inline src/development/troubleshoot.md]: My previous comment hasn't been fixed there?

File: src/development/troubleshoot.md [hunk 12]
```diff
@@ -161,3 +135,74 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+Related documentation: [Build and deploy hooks](https://docs.platform.sh/configuration/app/build.html#hooks)
+
+### Cron jobs
+
+In order to switch containers, we need to wait for all the current running tasks to finish. Long running cron jobs prevent the container from switching.
+
+First, make sure your custom cron jobs execution times are low and that they are running properly. Second, investigate your application cron jobs, as they can invoke other services in unexpected ways, which may increase execution time.
+
+**note**
+Drupal `drush core-cron` run installed module’s cron task. Those can be, for example; evicting invalid cache, updating database records, regenerating assets. Be sure to constantly benchmark the `drush core-cron` command in all your environments, as it is a common source of performance issues.
```
Comment (ralt) [inline src/development/troubleshoot.md]: Weird apostrophe :)

File: src/development/troubleshoot.md (hunk 11 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: I changed "command in your container" to "command in your environment".

Comment (gnufred) [reply src/development/troubleshoot.md]: If that's now what you mean then I don't understand the comment.

File: src/development/troubleshoot.md [hunk 13]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+## Stuck build (or deployment)
+
+If you see a build or deployment running longer than expected, that may be one of the following cases:
+
+1. The build is blocked by a process in your build hook.
+2. The deployment is blocked by a long running process in your deploy hook.
+3. The deployment is blocked by a long running cron job in the environment.
+4. The deployment is blocked by a long running cron job in the parent environment.
+
+To determine if your environment is being stuck in the build or the deployment, you should look at the build log available on the UI. When you see a log line like below, your application is built and the deployment is stuck:
```
Comment (Crell) [inline src/development/troubleshoot.md]: Minor nitpick: Remove "you should".  Either change to "you can" or omit entirely.  "You should" feels too much like an accusatory command here.  (Entirely subjective, I know.)

File: src/development/troubleshoot.md [hunk 14]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+Re-deploying environment w6ikvtghgyuty-drupal8-b3dsina.
+```
+
+Here are the things you should do when you see your build or deployment is stuck:
+
+For a blocked _build_ (when you don't find the `Re-deployment environment ...` line), you have to create a [support ticket](https://platform.sh/support) to let us kill it.
+
+When a _deployment_ is blocked, you should try the following:
+
+1. Use [SSH](/development/access-site.md) to connect to your environment. Find any long-running cron jobs on the environment by running `ps afx`. Once you have identified the long running process on the environment, kill it.
```
Comment (Crell) [inline src/development/troubleshoot.md]: Since we're expanding this section anyway, let's include briefly how one goes about killing a process.  Just adding `kill <PID>` with a half-sentence description should be sufficient.

File: src/development/troubleshoot.md [hunk 15]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+2. If you're performing "Sync", "Merge", or "Activate" on an environment and the process is stuck, use [SSH](/development/access-site.md) to connect to the parent environment and identify any long running cron jobs with `ps afx`. Kill the job(s) if you see any.
+
+Stuck builds should automaticly be killed after one hour. If they are stuck for a longer period please create [support ticket](https://platform.sh/support).
```
Comment (Crell) [inline src/development/troubleshoot.md]: Is this true anywhere other than Australia yet?  I didn't think that's been deployed on US or EU.

File: src/development/troubleshoot.md [hunk 16]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+### Check php.access.log
+
+The php.access.log contains the page executions by php-fpm. It also includes the execution time and peak memory usage of each request. This can be used to find pages and scripts that use too much memory or time to finish.
+
+Show 10 slowest page loads in the last 1000 requests: `tail -n 1000 php.access.log | sort -n -k 4 | tail`
+
+### Build and deploy hooks
+
+Hooks are almost always the cause of long build time. If they run into problem they can cause the build to fail or hang indefinitely.
```
Comment (Crell) [inline src/development/troubleshoot.md]: I'd say frequently here.  I've seen projects take upwards of 10 min just to download a zillion Composer and NPM dependencies before the build hook even enters the picture.

We should also split the build and deploy hooks to separate subsections, as the way you debug and address them is different as is the symptom of not doing so.  (Slow build is annoying, slow deploy hooks cause downtime.)

File: src/development/troubleshoot.md [hunk 17]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+### Check for errors in the logs
+
+Invisible errors during the build and deploy phase can cause increased wait times, failed builds and other problems. Investigating each log and fixing errors is essential.
+
+Related documentation: [Accessing logs](https://docs.platform.sh/development/logs.html#accessing-logs)
+
+### Check php.access.log
+
+The php.access.log contains the page executions by php-fpm. It also includes the execution time and peak memory usage of each request. This can be used to find pages and scripts that use too much memory or time to finish.
```
Comment (Crell) [inline src/development/troubleshoot.md]: This should add a caveat about "if you're using PHP".  Presumably that file won't exist otherwise.

Comment (Crell) [reply src/development/troubleshoot.md]: Also, PHP-FPM is capitalized.

File: src/development/troubleshoot.md [hunk 18]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Show 10 slowest page loads in the last 1000 requests: `tail -n 1000 php.access.log | sort -n -k 4 | tail`
```
Comment (Crell) [inline src/development/troubleshoot.md]: Nifty!  I like.

File: src/development/troubleshoot.md [hunk 19]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+You should test each hook and deploy command in your local development environment.
+
+Furthermore, you can test your hooks with those linux commands to help figure out any problems
+
+```
+time $cmd # Print execution time
+strace -T $cmd # Print a system call report
+```
+
+Related documentation: [Build and deploy hooks](https://docs.platform.sh/configuration/app/build.html#hooks)
```
Comment (Crell) [inline src/development/troubleshoot.md]: Somewhere in this section we should note that rerunning deploy hooks on an active site is *usually* safe, at least from our end, so debugging it there is fine.  (Modulo cache clearing and the impact of that, etc., which we should mention.)

Build hooks have to be debugged locally, but you can also see how long they take on push because the log is streamed in real-time.  So you can see if composer just takes forever or npm does or something else.

(This is why build and deploy should get their own sections.)

File: src/development/troubleshoot.md [hunk 20]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+### Cron jobs
+
+In order to switch containers, we need to wait for all the current running tasks to finish. Long running cron jobs prevent the container from switching.
```
Comment (Crell) [inline src/development/troubleshoot.md]: Can we reword this to not need to refer to Platform.sh directly?  We should also explain why we do that.  (Data integrity.)

File: src/development/troubleshoot.md [hunk 21]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+First, make sure your custom cron jobs execution times are low and that they are running properly. Second, investigate your application cron jobs, as they can invoke other services in unexpected ways, which may increase execution time.
+
+**note**
+Drupal `drush core-cron` run installed module's cron task. Those can be, for example; evicting invalid cache, updating database records, regenerating assets. Be sure to constantly benchmark the `drush core-cron` command in all your environments, as it is a common source of performance issues.
```
Comment (Crell) [inline src/development/troubleshoot.md]: "Drupal's drush core-cron runs"

Comment (Crell) [reply src/development/troubleshoot.md]: Also, don't benchmark it constantly.  Benchmark it frequently.  Constant benchmarking is a little excessive. :smile:

File: src/development/troubleshoot.md [hunk 22]
```diff
@@ -161,3 +135,75 @@ where `<project_id>` is the random-character ID of the project.  That can be fou
...
+
+Related documentation: [Cron and scheduled tasks](https://docs.platform.sh/configuration/app/cron.html#cron-jobs)
```
Comment (Crell) [inline src/development/troubleshoot.md]: Love the deep links!

File: src/development/troubleshoot.md (hunk 15 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: It's not but @ralt said to write this down here so here it is!

File: src/development/troubleshoot.md (hunk 16 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: I don't know enough about the inner working to be able to improve this section. Can you add what's needed to the guide?

File: src/development/troubleshoot.md (hunk 19 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: @ralt Mentioned it was impossible due to the fact that the environment is read-only, so it needs to be tested locally.

File: src/development/troubleshoot.md (hunk 20 above)
Comment (gnufred) [reply src/development/troubleshoot.md]: This is the best I can do. Can you fill the gap?

File: src/development/troubleshoot.md (hunk 15 above)
Comment (Crell) [reply src/development/troubleshoot.md]: @ralt How long before US and EU have build time limits?  No sense being misleading for the majority of our customer base.

**Comment by Crell** [comment]
@gnufred I think you said you had new commits to push? I don't want to tweak myself unless I know have your latest branch.

**Comment by gnufred** [comment]
@Crell I just pushed the latest changes.
